In [1]:
import pandas as pd

In [2]:
s = pd.Series([1,2,3,4,5])

In [3]:
s

0    1
1    2
2    3
3    4
4    5
dtype: int64

In [4]:
s.head(3)

0    1
1    2
2    3
dtype: int64

ufo = pd.read_table("http://bit.ly/uforeports", sep=",")

In [7]:
ufo.head()

City Colors Reported Shape Reported State             Time
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00
3               Abilene             NaN           DISK    KS   6/1/1931 13:00
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00

In [8]:
ufo['State']

0        NY
1        NJ
2        CO
3        KS
4        NY
5        ND
6        CA
7        MI
8        AK
9        OR
10       CA
11       AL
12       SC
13       IA
14       MI
15       CA
16       CA
17       GA
18       TN
19       AK
20       NE
21       LA
22       LA
23       KY
24       WV
25       CA
26       WV
27       NM
28       NM
29       UT
         ..
18211    MA
18212    CA
18213    CA
18214    TX
18215    TX
18216    CA
18217    CO
18218    TX
18219    CA
18220    CA
18221    NH
18222    PA
18223    SC
18224    OK
18225    CA
18226    CA
18227    CA
18228    TX
18229    IL
18230    CA
18231    CA
18232    WI
18233    AK
18234    CA
18235    AZ
18236    IL
18237    IA
18238    WI
18239    WI
18240    FL
Name: State, Length: 18241, dtype: object

In [9]:
ufo.State

0        NY
1        NJ
2        CO
3        KS
4        NY
5        ND
6        CA
7        MI
8        AK
9        OR
10       CA
11       AL
12       SC
13       IA
14       MI
15       CA
16       CA
17       GA
18       TN
19       AK
20       NE
21       LA
22       LA
23       KY
24       WV
25       CA
26       WV
27       NM
28       NM
29       UT
         ..
18211    MA
18212    CA
18213    CA
18214    TX
18215    TX
18216    CA
18217    CO
18218    TX
18219    CA
18220    CA
18221    NH
18222    PA
18223    SC
18224    OK
18225    CA
18226    CA
18227    CA
18228    TX
18229    IL
18230    CA
18231    CA
18232    WI
18233    AK
18234    CA
18235    AZ
18236    IL
18237    IA
18238    WI
18239    WI
18240    FL
Name: State, Length: 18241, dtype: object

In [10]:
ufo.Location = ufo.City + "," + ufo.State

In [11]:
ufo.head()

City Colors Reported Shape Reported State             Time
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00
3               Abilene             NaN           DISK    KS   6/1/1931 13:00
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00

In [12]:
ufo.Location

0                      Ithaca,NY
1                 Willingboro,NJ
2                     Holyoke,CO
3                     Abilene,KS
4        New York Worlds Fair,NY
5                 Valley City,ND
6                 Crater Lake,CA
7                        Alma,MI
8                     Eklutna,AK
9                     Hubbard,OR
10                    Fontana,CA
11                   Waterloo,AL
12                     Belton,SC
13                     Keokuk,IA
14                  Ludington,MI
15                Forest Home,CA
16                Los Angeles,CA
17                  Hapeville,GA
18                     Oneida,TN
19                 Bering Sea,AK
20                   Nebraska,NE
21                           NaN
22                           NaN
23                  Owensboro,KY
24                 Wilderness,WV
25                  San Diego,CA
26                 Wilderness,WV
27                     Clovis,NM
28                 Los Alamos,NM
29               Ft. Duschene,UT
          

In [13]:
ufo['Location'] = ufo.City + "," + ufo.State

In [14]:
ufo.Location

0                      Ithaca,NY
1                 Willingboro,NJ
2                     Holyoke,CO
3                     Abilene,KS
4        New York Worlds Fair,NY
5                 Valley City,ND
6                 Crater Lake,CA
7                        Alma,MI
8                     Eklutna,AK
9                     Hubbard,OR
10                    Fontana,CA
11                   Waterloo,AL
12                     Belton,SC
13                     Keokuk,IA
14                  Ludington,MI
15                Forest Home,CA
16                Los Angeles,CA
17                  Hapeville,GA
18                     Oneida,TN
19                 Bering Sea,AK
20                   Nebraska,NE
21                           NaN
22                           NaN
23                  Owensboro,KY
24                 Wilderness,WV
25                  San Diego,CA
26                 Wilderness,WV
27                     Clovis,NM
28                 Los Alamos,NM
29               Ft. Duschene,UT
          

In [15]:
ufo.head()

City Colors Reported Shape Reported State             Time  \
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00   
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00   
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00   
3               Abilene             NaN           DISK    KS   6/1/1931 13:00   
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00   

                  Location  
0                Ithaca,NY  
1           Willingboro,NJ  
2               Holyoke,CO  
3               Abilene,KS  
4  New York Worlds Fair,NY

# Check that when we just said ufo.Location = ufo.City + "," + ufo.State, it worked and got printed using ufo.Location but when we printed head() then this data did not come through. When we did using ufo['Location'] = ufo.City + "," + ufo.State, then it got displayed in the head. TODO - What is the difference here.

In [16]:
ufo.shape
ufo.dtypes

City               object
Colors Reported    object
Shape Reported     object
State              object
Time               object
Location           object
dtype: object

In [17]:
ufo.shape

(18241, 6)

# All of the attributed are displayed as an object, so what kind of object we mean here??

In [18]:
ufo.drop('Colors Reported', axis=1, inplace=True)

In [19]:
ufo.head()

City Shape Reported State             Time  \
0                Ithaca       TRIANGLE    NY   6/1/1930 22:00   
1           Willingboro          OTHER    NJ  6/30/1930 20:00   
2               Holyoke           OVAL    CO  2/15/1931 14:00   
3               Abilene           DISK    KS   6/1/1931 13:00   
4  New York Worlds Fair          LIGHT    NY  4/18/1933 19:00   

                  Location  
0                Ithaca,NY  
1           Willingboro,NJ  
2               Holyoke,CO  
3               Abilene,KS  
4  New York Worlds Fair,NY

# If we need to drop multiple columns, then simply provide the column names in a list such as ['State', 'Time']

In [ ]:
#Let us bring back the original data

In [20]:
ufo = pd.read_table("http://bit.ly/uforeports", sep=",")
ufo.head()

City Colors Reported Shape Reported State             Time
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00
3               Abilene             NaN           DISK    KS   6/1/1931 13:00
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00

In [21]:
ufo.State.sort_values(ascending=False)
ufo.head()

City Colors Reported Shape Reported State             Time
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00
3               Abilene             NaN           DISK    KS   6/1/1931 13:00
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00

In [22]:
ufo.State.sort_values(ascending=False).head()

12079    WY
11490    WY
11333    WY
4866     WY
3326     WY
Name: State, dtype: object

# See the difference when I typed ufo.head() separately? It did not print the change, but when we typed head() with sorting, it printed the change. This seems to say that the sorting is NOT in place. TODO - confirm this point and dig more into that.

# TODO - though WY is same, we see another column with different values like 12079 etc and thats not sorted, whats these values??

In [24]:
ufo.sort_values('City').head()

City Colors Reported Shape Reported State              Time
1761   Abbeville             NaN           DISK    SC   12/10/1968 0:30
4553    Aberdeen             NaN       CYLINDER    WA   6/15/1981 22:00
16167   Aberdeen             NaN        VARIOUS    OH    3/29/2000 3:00
14703   Aberdeen             NaN       TRIANGLE    WA   9/30/1999 21:00
389     Aberdeen          ORANGE         CIRCLE    SD  11/15/1956 18:30

# TODO - What is the difference between ufo.sort_values('City').head() and ufo.State.sort_values(ascending=False)???

# NOTE - we can sort by two columns also. Note that multiple attributes are always passed in a list here.

In [25]:
ufo.sort_values(['City', 'State']).head()

City Colors Reported Shape Reported State              Time
1761   Abbeville             NaN           DISK    SC   12/10/1968 0:30
2297    Aberdeen             NaN       TRIANGLE    MD    8/18/1972 1:30
9404    Aberdeen             NaN           DISK    MD   6/15/1996 13:30
16167   Aberdeen             NaN        VARIOUS    OH    3/29/2000 3:00
389     Aberdeen          ORANGE         CIRCLE    SD  11/15/1956 18:30